pkg/util/vsphere/watcher/watcher.go

```diff
@@ -0,0 +1,672 @@
+					if ov := aov.OptionValue[j].GetOptionValue(); ov != nil {
+
+						// Get the namespace and name of the VM from the changes
+						// if they are present there.
+						if ov.Key == extraConfigNamespacedNameKey {
+							if namespace == "" || name == "" {
+								if s, ok := ov.Value.(string); ok {
+									namespace, name = namespacedNameFromString(s)
+								}
+							}
+						}
+
+						// Note if the key is ignored.
+						if _, ok := w.ignoredExtraConfigKeys[ov.Key]; ok {
+							ignoredOptVals++
+						}
+					}
+				}
+
+				// The change to extraConfig is ignored if the number of ignored
+				// option values is the same as the number of total option vals.
+				ignore = ignoredOptVals == len(aov.OptionValue)
+			}
+		}
+		if ignore {
+			ignoredChanges++
+		}
+	}
+	if ignoredChanges == len(update.changes) {
+		return nil
+	}
+
+	if w.lookupNamespacedName != nil {
+		if namespace == "" || name == "" || update.kind == vimtypes.ObjectUpdateKindEnter {
+			namespace, name, verified = w.lookupNamespacedName(ctx, obj)
+		}
+	}
+
+	if update.kind == vimtypes.ObjectUpdateKindEnter && verified {
+		// The behavior of Controller-Runtime to sync all objects upon startup
+		// will cause *existing* VMs to be reconciled. Therefore, do not emit a
+		// result when the object is entering the scope of the watcher and the
+		// corresponding Kubernetes object already exists with a matching
+		// status.uniqueID field.
+		return nil
+	}
+
+	if namespace == "" || name == "" {
+		rep, err := property.DefaultCollector(w.client).RetrieveProperties(
+			ctx,
+			vimtypes.RetrieveProperties{
+				SpecSet: []vimtypes.PropertyFilterSpec{
+					{
+						ObjectSet: []vimtypes.ObjectSpec{
+							{
+								Obj: obj,
+							},
+						},
+						PropSet: []vimtypes.PropertySpec{
+							{
+								Type: "VirtualMachine",
+								PathSet: []string{
+									extraConfigNamespaceNamePropPath,
+								},
+							},
+						},
+					},
+				},
+			})
+		if err != nil {
+			return err
+		}
+		namespace, name = namespacedNameFromObjContent(rep.Returnval)
+	}
+
+	if namespace != "" && name != "" {
+		r := Result{
+			Namespace: namespace,
+			Name:      name,
+			Ref:       obj,
+			Verified:  verified,
+		}
+
+		logger.Info("sending result", "result", r)
+
+		go func(r Result) {
+			w.chanResult <- r
+		}(r)
+	}
+
+	return nil
+}
+
+// Add starts watching a container to which VirtualMachine resources may belong,
+// such as a Folder, Cluster, ResourcePool, etc.
+func Add(ctx context.Context, ref moRef) {
+	go func() {
+		fromContext(ctx).add <- ref
+	}()
+}
+
+// Remove stops watching a container to which VirtualMachine resources may
+// belong, such as a Folder, Cluster, ResourcePool, etc.
+func Remove(ctx context.Context, ref moRef) {
+	go func() {
+		fromContext(ctx).rem <- ref
+	}()
+}
+
+func (w *Watcher) add(ctx context.Context, ref moRef) error {
+	w.mu.Lock()
+	defer w.mu.Unlock()
+
+	if w.pc == nil {
+		return nil
+	}
+
```

Comment:
Nit: Probably add a comment here 
// return early if containerView is already present

Reply:
It returns early if the watcher is closed, but sure, will add comment.

Reply:
was referring to L483 : 
`if _, ok := w.cv[ref]; ok {`